src/App/DataSource/DataSourceManager.php

```diff
@@ -104,25 +149,130 @@ public function getSourceForType($type)
         return false;
     }
 
-    public function registerRoutes()
+    /**
+     * Register routes for callback sources
+     */
+    public function registerRoutes(Router $router)
     {
-        foreach ($this->getEnabledSources() as $source) {
-            if (!($source instanceof CallbackDataSource)) {
-                // Data source doesn't implement callbacks
-                continue;
+        foreach ($this->sources as $name => $class) {
+            if (in_array(CallbackDataSource::class, class_implements($class))) {
+                $class::registerRoutes($router);
             }
-
-            $source->registerRoutes($this->router);
         }
     }
 
-    public function setStorage($storage)
+    /**
+     * Set data source storage
+     * @param DataSourceStorage $storage
+     */
+    public function setStorage(DataSourceStorage $storage)
     {
         $this->storage = $storage;
     }
 
-    public function getStorage()
+    /**
+     * Get data source storage
+     * @return DataSourceStorage
+     */
+    public function getStorage() : DataSourceStorage
     {
         return $this->storage;
     }
+
+    /**
+     * Get config for data source
+     * @param  string $name
+     * @return array
+     */
+    protected function getConfig(string $name) : array
+    {
+        $config = Cache::remember('config.data-provider', self::CACHE_LIFETIME, function () {
+            return $this->configRepo->get('data-provider')->asArray();
+        });
+
+        return $config[$name] ?? [];
+    }
+
+    /**
+     * Resolve data source class
+     * @param  string $name
+     * @return DataSource
+     */
+    protected function resolve(string $name) : DataSource
+    {
+        $config = $this->getConfig($name);
+
+        $driverMethod = 'create'.ucfirst($name).'Source';
+
+        if (method_exists($this, $driverMethod)) {
+            return $this->loadedSources[$name] = $this->{$driverMethod}($config);
+        } else {
+            throw new InvalidArgumentException("Source [{$name}] is not supported.");
+        }
+    }
+
+    /**
+     * Wipe resolved source instances
+     */
+    public function clearResolvedSources()
+    {
+        $this->loadedSources = [];
+    }
+
+    protected function createSmssyncSource(array $config)
+    {
+        return new SMSSync\SMSSync($config);
+    }
+
+    protected function createEmailSource(array $config)
+    {
+        return new Email\Email(
+            $config,
+            app('mailer'),
+            app(\Ushahidi\Core\Entity\MessageRepository::class)
+        );
+    }
+
+    protected function createOutgoingemailSource(array $config)
```

Comment:
nitpick: if we're calling it OutgoingEmail elsewhere, can we adjust the camel case here?

Reply:
Not easily because we generate the method name in `resolve` with `$driverMethod = 'create'.ucfirst($name).'Source';`. We'd have to rely on `$name` already being camelcase